File: src/js/modules/http/http_ssl_context.cpp
#include <OSS/JSON/elements.h>

#include "OSS/JS/JSPlugin.h"
#include "OSS/UTL/CoreUtils.h"
#include "OSS/UTL/Logger.h"
#include "OSS/JSON/Json.h"
#include "OSS/JS/modules/Async.h"
#include "OSS/JS/modules/HttpSSLContextObject.h"

using OSS::JS::ObjectWrap;

JS_CLASS_INTERFACE(HttpSSLContextObject, "HttpSSLContext") 
{
  JS_CLASS_METHOD_DEFINE(HttpSSLContextObject, "setUsage", setUsage);
  JS_CLASS_METHOD_DEFINE(HttpSSLContextObject, "getUsage", getUsage);
  JS_CLASS_METHOD_DEFINE(HttpSSLContextObject, "setPrivateKeyFile", setPrivateKeyFile);
  JS_CLASS_METHOD_DEFINE(HttpSSLContextObject, "getPrivateKeyFile", getPrivateKeyFile);
  JS_CLASS_METHOD_DEFINE(HttpSSLContextObject, "setCertificateKeyFile", setCertificateKeyFile);
  JS_CLASS_METHOD_DEFINE(HttpSSLContextObject, "getCertificateKeyFile", getCertificateKeyFile);
  JS_CLASS_METHOD_DEFINE(HttpSSLContextObject, "setCaLocation", setCaLocation);
  JS_CLASS_METHOD_DEFINE(HttpSSLContextObject, "getCaLocation", getCaLocation);
  JS_CLASS_METHOD_DEFINE(HttpSSLContextObject, "setVerificationMode", setVerificationMode);
  JS_CLASS_METHOD_DEFINE(HttpSSLContextObject, "getVerificationMode", getVerificationMode);
  JS_CLASS_METHOD_DEFINE(HttpSSLContextObject, "setVerificationDepth", setVerificationDepth);
  JS_CLASS_METHOD_DEFINE(HttpSSLContextObject, "getVerificationDepth", getVerificationDepth);
  JS_CLASS_METHOD_DEFINE(HttpSSLContextObject, "setLoadDefaultCa", setLoadDefaultCa);
  JS_CLASS_METHOD_DEFINE(HttpSSLContextObject, "getLoadDefaultCa", getLoadDefaultCa);
  JS_CLASS_METHOD_DEFINE(HttpSSLContextObject, "setCipherList", setCipherList);
  JS_CLASS_METHOD_DEFINE(HttpSSLContextObject, "getCipherList", getCipherList);
  JS_CLASS_METHOD_DEFINE(HttpSSLContextObject, "setPassphraseHandlerId", setPassphraseHandlerId);
  JS_CLASS_METHOD_DEFINE(HttpSSLContextObject, "setInvalidCertificateHandlerId", setInvalidCertificateHandlerId);
  
  JS_CLASS_METHOD_DEFINE(HttpSSLContextObject, "registerContext", registerContext);
  
  JS_CLASS_INTERFACE_END(HttpSSLContextObject);
}

class CertificateValidator : public HttpSSLContextObject::InvalidCertificateHandler
{
public:
  HttpSSLContextObject* _pContext;
  
  CertificateValidator(HttpSSLContextObject* pContext) :
    HttpSSLContextObject::InvalidCertificateHandler(pContext->_usage == HttpSSLContextObject::Context::SERVER_USE ? true : false)
  {
    _pContext = pContext;
  }
  
  ~CertificateValidator()
  {
  }
  
  void onInvalidCertificate(const void* pSender, HttpSSLContextObject::VerificationErrorArgs& errorCert)
  {
    OSS::JSON::Object request, certificate, result;
    
    request["method"] = OSS::JSON::String(_pContext->_invalidCertHandlerId);
    const Poco::Net::X509Certificate& cert = errorCert.certificate();
    
    certificate["errorMessage"] = OSS::JSON::String(errorCert.errorMessage());
    certificate["commonName"] = OSS::JSON::String(cert.commonName());
    certificate["issuerName"] = OSS::JSON::String(cert.issuerName());
    certificate["subjectName"] = OSS::JSON::String(cert.subjectName());
    
    std::set<std::string> domainNames;
    std::string commonaName;
    cert.extractNames(commonaName, domainNames);
    if (!domainNames.empty())
    {
      OSS::JSON::Array aliases;
      std::size_t index = 0;
      for (std::set<std::string>::iterator iter = domainNames.begin(); iter != domainNames.end(); iter++)
      {
        aliases[index] = OSS::JSON::String(*iter);
        index++;
        certificate["domains"] = aliases;
      }
    }
    
    request["arguments"] = certificate;
    if (Async::json_execute_promise(request, result))
    {
      OSS::JSON::Object::iterator iter = result.Find("result");
      if (iter != result.End())
      {
        OSS::JSON::Object& resultValue = iter->element;
        OSS::JSON::Object::iterator allowIter = resultValue.Find("allow");
        if (allowIter != resultValue.End())
        {
          OSS::JSON::Boolean& allow = allowIter->element;
          errorCert.setIgnoreError(allow.Value());
        }
      }
    }
  }
};

class PassPhraseValidator : public HttpSSLContextObject::PrivateKeyPassphraseHandler
{
public:
  HttpSSLContextObject* _pContext;
  
  PassPhraseValidator(HttpSSLContextObject* pContext) :
    HttpSSLContextObject::PrivateKeyPassphraseHandler(pContext->_usage == HttpSSLContextObject::Context::SERVER_USE ? true : false)
  {
    _pContext = pContext;
  }
  
  ~PassPhraseValidator()
  {
  }
  
  void onPrivateKeyRequested(const void* pSender, std::string& privateKey)
  {
    OSS::JSON::Object request, result;
    request["method"] = OSS::JSON::String(_pContext->_passPhraseHandlerId);
    if (Async::json_execute_promise(request, result))
    {
   
      OSS::JSON::Object::iterator iter = result.Find("result");
      if (iter != result.End())
      {
        OSS::JSON::Object& resultValue = iter->element;
        OSS::JSON::Object::iterator passwordIter = resultValue.Find("password");
        if (passwordIter != resultValue.End())
        {
          OSS::JSON::String& password = passwordIter->element;
          privateKey = password.Value();
        }
      }
    }
  }
};

HttpSSLContextObject::HttpSSLContextObject()
{
  _usage = Context::CLIENT_USE;
  _cipherList = "ALL:!ADH:!LOW:!EXP:!MD5:@STRENGTH";
  _verificationMode = Context::VERIFY_RELAXED;
  _verificationDepth = 9;
  _loadDefaultCAs = false;
}

HttpSSLContextObject::~HttpSSLContextObject()
{
}

JS_CONSTRUCTOR_IMPL(HttpSSLContextObject)
{
  js_enter_scope();
  HttpSSLContextObject* pObject = new HttpSSLContextObject();
  pObject->Wrap(js_method_arg_self());
  return js_method_arg_self();
}

JS_METHOD_IMPL(HttpSSLContextObject::setUsage)
{
  js_enter_scope();
  HttpSSLContextObject* self = js_method_arg_unwrap_self(HttpSSLContextObject);
  js_method_arg_declare_int32(usage, 0);
  self->_usage = (Context::Usage)usage;
  return JSUndefined();
}

JS_METHOD_IMPL(HttpSSLContextObject::getUsage)
{
  js_enter_scope();
  HttpSSLContextObject* self = js_method_arg_unwrap_self(HttpSSLContextObject);
  return JSInt32(self->_usage);
}

JS_METHOD_IMPL(HttpSSLContextObject::setPrivateKeyFile)
{
  js_enter_scope();
  HttpSSLContextObject* self = js_method_arg_unwrap_self(HttpSSLContextObject);
  js_method_arg_declare_string(privateKeyFile, 0);
  self->_privateKeyFile = privateKeyFile;
  return JSUndefined();
}
JS_METHOD_IMPL(HttpSSLContextObject::getPrivateKeyFile)
{
  js_enter_scope();
  HttpSSLContextObject* self = js_method_arg_unwrap_self(HttpSSLContextObject);
  return JSString(self->_privateKeyFile);
}
JS_METHOD_IMPL(HttpSSLContextObject::setCertificateKeyFile)
{
  js_enter_scope();
  HttpSSLContextObject* self = js_method_arg_unwrap_self(HttpSSLContextObject);
  js_method_arg_declare_string(certificateFile, 0);
  self->_certificateFile = certificateFile;
  return JSUndefined();
}
JS_METHOD_IMPL(HttpSSLContextObject::getCertificateKeyFile)
{
  js_enter_scope();
  HttpSSLContextObject* self = js_method_arg_unwrap_self(HttpSSLContextObject);
  return JSString(self->_certificateFile);
}

JS_METHOD_IMPL(HttpSSLContextObject::setCaLocation)
{
  js_enter_scope();
  HttpSSLContextObject* self = js_method_arg_unwrap_self(HttpSSLContextObject);
  js_method_arg_declare_string(caLocation, 0);
  self->_caLocation = caLocation;
  return JSUndefined();
}
JS_METHOD_IMPL(HttpSSLContextObject::getCaLocation)
{
  js_enter_scope();
  HttpSSLContextObject* self = js_method_arg_unwrap_self(HttpSSLContextObject);
  return JSString(self->_caLocation);
}

JS_METHOD_IMPL(HttpSSLContextObject::setCipherList)
{
  js_enter_scope();
  HttpSSLContextObject* self = js_method_arg_unwrap_self(HttpSSLContextObject);
  js_method_arg_declare_string(cipherList, 0);
  self->_cipherList = cipherList;
  return JSUndefined();
}

JS_METHOD_IMPL(HttpSSLContextObject::getCipherList)
{
  js_enter_scope();
  HttpSSLContextObject* self = js_method_arg_unwrap_self(HttpSSLContextObject);
  return JSString(self->_cipherList);
}

JS_METHOD_IMPL(HttpSSLContextObject::setVerificationMode)
{
  js_enter_scope();
  HttpSSLContextObject* self = js_method_arg_unwrap_self(HttpSSLContextObject);
  js_method_arg_declare_int32(verificationMode, 0);
  self->_verificationMode = (Context::VerificationMode)verificationMode;
  return JSUndefined();
}

JS_METHOD_IMPL(HttpSSLContextObject::getVerificationMode)
{
  js_enter_scope();
  HttpSSLContextObject* self = js_method_arg_unwrap_self(HttpSSLContextObject);
  return JSInt32(self->_verificationMode);
}

JS_METHOD_IMPL(HttpSSLContextObject::setVerificationDepth)
{
  js_enter_scope();
  HttpSSLContextObject* self = js_method_arg_unwrap_self(HttpSSLContextObject);
  js_method_arg_declare_int32(verificationDepth, 0);
  self->_verificationDepth = verificationDepth;
  return JSUndefined();
}

JS_METHOD_IMPL(HttpSSLContextObject::getVerificationDepth)
{
  js_enter_scope();
  HttpSSLContextObject* self = js_method_arg_unwrap_self(HttpSSLContextObject);
  return JSInt32(self->_verificationDepth);
}

JS_METHOD_IMPL(HttpSSLContextObject::setLoadDefaultCa)
{
  js_enter_scope();
  HttpSSLContextObject* self = js_method_arg_unwrap_self(HttpSSLContextObject);
  js_method_arg_declare_bool(loadDefaultCa, 0);
  self->_loadDefaultCAs = loadDefaultCa;
  return JSUndefined();
}

JS_METHOD_IMPL(HttpSSLContextObject::getLoadDefaultCa)
{
  js_enter_scope();
  HttpSSLContextObject* self = js_method_arg_unwrap_self(HttpSSLContextObject);
  return JSBoolean(self->_loadDefaultCAs);
}

JS_METHOD_IMPL(HttpSSLContextObject::setInvalidCertificateHandlerId)
{
  js_enter_scope();
  HttpSSLContextObject* self = js_method_arg_unwrap_self(HttpSSLContextObject);
  js_method_arg_declare_string(invalidCertHandlerId, 0);
  self->_invalidCertHandlerId = invalidCertHandlerId;
  return JSUndefined();
}

JS_METHOD_IMPL(HttpSSLContextObject::setPassphraseHandlerId)
{
  js_enter_scope();
  HttpSSLContextObject* self = js_method_arg_unwrap_self(HttpSSLContextObject);
  js_method_arg_declare_string(passPhraseHandlerId, 0);
  self->_passPhraseHandlerId = passPhraseHandlerId;
  return JSUndefined();
}

JS_METHOD_IMPL(HttpSSLContextObject::registerContext)
{
  js_enter_scope();
  HttpSSLContextObject* self = js_method_arg_unwrap_self(HttpSSLContextObject);
  
  if (self->_pContext)
  {
    js_throw("Context Already Registered");
  }
  
  try
  {
    self->_pPassPhraseHandler = new PassPhraseValidator(self);
    self->_pInvalidCertHandler = new CertificateValidator(self);
    self->_pContext = new Context(
      self->_usage, 
      self->_privateKeyFile, 
      self->_certificateFile, 
      self->_caLocation, 
      self->_verificationMode, 
      self->_verificationDepth, 
      self->_loadDefaultCAs, 
      self->_cipherList);

    if (self->_usage == HttpSSLContextObject::Context::SERVER_USE)
    {
      SSLManager::instance().initializeServer(self->_pPassPhraseHandler, self->_pInvalidCertHandler, self->_pContext);
    }
    else
    {
      SSLManager::instance().initializeClient(self->_pPassPhraseHandler, self->_pInvalidCertHandler, self->_pContext);
    }
  }
  catch(const Poco::Exception& e)
  {
    js_throw(e.message().c_str());
  }
  return JSUndefined();
}

JS_EXPORTS_INIT()
{
  js_enter_scope();
  js_export_class(HttpSSLContextObject);
  
  js_export_int32("CLIENT_USE", HttpSSLContextObject::Context::CLIENT_USE);
  js_export_int32("SERVER_USE", HttpSSLContextObject::Context::SERVER_USE);
  js_export_int32("VERIFY_NONE", HttpSSLContextObject::Context::VERIFY_NONE);
  js_export_int32("VERIFY_RELAXED", HttpSSLContextObject::Context::VERIFY_RELAXED);
  js_export_int32("VERIFY_STRICT", HttpSSLContextObject::Context::VERIFY_STRICT);
  js_export_int32("VERIFY_ONCE", HttpSSLContextObject::Context::VERIFY_ONCE);
  
  js_export_finalize(); 
}

JS_REGISTER_MODULE(JSHttpSSLContext);

File: src/js/modules/http/http_ssl_context.js
"use-strict";

const _http_ssl_context = require("./_http_ssl_context.jso");
const SSLContext = _http_ssl_context.HttpSSLContext;
const EventEmitter = async.EventEmitter;
const isolate = require("isolate");

///   * usage specifies whether the context is used by a client or server.
///   * privateKeyFile contains the path to the private key file used for encryption.
///     Can be empty if no private key file is used.
///   * certificateFile contains the path to the certificate file (in PEM format).
///     If the private key and the certificate are stored in the same file, this
///     can be empty if privateKeyFile is given.
///   * caLocation contains the path to the file or directory containing the
///     CA/root certificates. Can be empty if the OpenSSL builtin CA certificates
///     are used (see loadDefaultCAs).
///   * verificationMode specifies whether and how peer certificates are validated.
///   * verificationDepth sets the upper limit for verification chain sizes. Verification
///     will fail if a certificate chain larger than this is encountered.
///   * loadDefaultCAs specifies wheter the builtin CA certificates from OpenSSL are used.
///   * cipherList specifies the supported ciphers in OpenSSL notation.
var HttpSSLContext = function(
  usage,
  privateKeyFile,
  certificateFile,
  caLocation,
  verificationMode,
  verificationDepth,
  loadDefaultCAs,
  cipherList)
{
  if (verificationMode === undefined)
  {
    verificationMode = HttpSSLContext.VERIFY_RELAXED;
  }
  if (verificationDepth === undefined)
  {
    verificationDepth = 9;
  }
  if (loadDefaultCAs === undefined)
  {
    loadDefaultCAs = false;
  }
  if (cipherList === undefined)
  {
    cipherList = "ALL:!ADH:!LOW:!EXP:!MD5:@STRENGTH";
  }
  
  var _context = new SSLContext();
  var _passphraseHandlerId = "HttpSSLContext_onPrivateKeyRequested_" + HttpSSLContext._INSTANCE_COUNTER;
  var _invalidCertificateHandlerId = "HttpSSLContext_onInvalidCertificate_" + HttpSSLContext._INSTANCE_COUNTER;
  ++HttpSSLContext._INSTANCE_COUNTER;
  
  var _passphraseHandlerCb = undefined;
  var passphraseHandlerCb = function()
  {
    var ret = _passphraseHandlerCb(certificate);
    if (ret)
    {
      return { result : { password : ret } }
    }
    else
    {
      return { result : {} }
    }
  }
  
  var _invalidCertificateHandlerCb = undefined;
  var invalidCertificateHandlerCb = function(certificate)
  {
    var ret = _invalidCertificateHandlerCb(certificate);
    return { result : { allow : !!ret } }
  }
  
  this.registerContext = function(passpharaseHandler, invalidCertificateHandler)
  {
    _context.setUsage(usage);
    _context.setPrivateKeyFile(privateKeyFile);
    _context.setCertificateKeyFile(certificateFile);
    _context.setCaLocation(caLocation);
    _context.setVerificationMode(verificationMode);
    _context.setVerificationDepth(verificationDepth);
    _context.setLoadDefaultCa(loadDefaultCAs);
    _context.setCipherList(cipherList);
    _context.setPassphraseHandlerId(_passphraseHandlerId);
    _context.setInvalidCertificateHandlerId(_invalidCertificateHandlerId);
    _context.registerContext();
    
    _passphraseHandlerCb = passpharaseHandler;
    _invalidCertificateHandlerCb = invalidCertificateHandler;
    
    isolate.on(_passphraseHandlerId, passphraseHandlerCb);
    isolate.on(_invalidCertificateHandlerId, invalidCertificateHandlerCb);
  }
  
  this.unregisterContext = function()
  {
    isolate.remove(_passphraseHandlerId);
    isolate.remove(_invalidCertificateHandlerId);
  }
}
HttpSSLContext._INSTANCE_COUNTER = 0;

exports.CLIENT_USE = HttpSSLContext.CLIENT_USE = _http_ssl_context.CLIENT_USE /// Context is used by a client.
exports.SERVER_USE = HttpSSLContext.SERVER_USE = _http_ssl_context.SERVER_USE  /// Context is used by a server.

exports.VERIFY_NONE = HttpSSLContext.VERIFY_NONE    = _http_ssl_context.VERIFY_NONE;
  /// Server: The server will not send a client certificate 
  /// request to the client, so the client will not send a certificate. 
  ///
  /// Client: If not using an anonymous cipher (by default disabled), 
  /// the server will send a certificate which will be checked, but
  /// the result of the check will be ignored.

exports.VERIFY_RELAXED = HttpSSLContext.VERIFY_RELAXED = _http_ssl_context.VERIFY_RELAXED;
  /// Server: The server sends a client certificate request to the 
  /// client. The certificate returned (if any) is checked. 
  /// If the verification process fails, the TLS/SSL handshake is 
  /// immediately terminated with an alert message containing the 
  /// reason for the verification failure. 
  ///
  /// Client: The server certificate is verified, if one is provided. 
  /// If the verification process fails, the TLS/SSL handshake is
  /// immediately terminated with an alert message containing the 
  /// reason for the verification failure. 

exports.VERIFY_STRICT = HttpSSLContext.VERIFY_STRICT  = _http_ssl_context.VERIFY_STRICT;
  /// Server: If the client did not return a certificate, the TLS/SSL 
  /// handshake is immediately terminated with a handshake failure
  /// alert. 
  ///
  /// Client: Same as VERIFY_RELAXED. 

exports.VERIFY_ONCE = HttpSSLContext.VERIFY_ONCE    = _http_ssl_context.VERIFY_ONCE;
  /// Server: Only request a client certificate on the initial 
  /// TLS/SSL handshake. Do not ask for a client certificate 
  /// again in case of a renegotiation.
  ///
  /// Client: Same as VERIFY_RELAXED.	

var _ssl_client_context = undefined;

exports.hasClientContext = function()
{
  return _ssl_client_context !== undefined;
}

exports.registerClientContext = function(
  privateKeyFile,
  passphraseFunction, 
  certificateFile,
  validateCertFunction,
  caLocation, 
  verificationMode, 
  verificationDepth, 
  loadDefaultCAs, 
  cipherList
)
{
  if (_ssl_client_context)
  {
    throw new Error("Client context already set");
  }
  _ssl_client_context = new HttpSSLContext(
    HttpSSLContext.CLIENT_USE,
    privateKeyFile,
    certificateFile,
    caLocation, 
    verificationMode, 
    verificationDepth, 
    loadDefaultCAs, 
    cipherList);
    _ssl_client_context.registerContext(passphraseFunction, validateCertFunction);
}

var _ssl_server_context = undefined;

exports.hasServerContext = function()
{
  return _ssl_server_context !== undefined;
}

exports.registerServerContext = function(
  privateKeyFile,
  passphraseFunction, 
  certificateFile,
  validateCertFunction,
  caLocation, 
  verificationMode, 
  verificationDepth, 
  loadDefaultCAs, 
  cipherList
)
{
  if (_ssl_server_context)
  {
    throw new Error("Server context already set");
  }
  
  _ssl_server_context = new HttpSSLContext(
    HttpSSLContext.SERVER_USE,
    privateKeyFile,
    certificateFile,
    caLocation, 
    verificationMode, 
    verificationDepth, 
    loadDefaultCAs, 
    cipherList);
    _ssl_client_context.registerContext(passphraseFunction, validateCertFunction);
}